### ISSUE TITLE
Add argument for Guppy "chunks_per_runner"

### ISSUE THREAD

**Issue by mshamash** [Psy-Fer/buttery-eel#2]
In testing buttery-eel with some BLOW5 data for basecalling, I realized that there is no "--chunks_per_runner" argument that could be set and passed through to Guppy.
Is this something that could be easily added to a future release? I had originally tuned Guppy performance with the chunks_per_runner parameter, as opposed to changing chunk_size or max_queued_reads, and being able to set that to the same as my previous runs would allow for direct comparison of basecalling times between buttery-eel and Guppy.
On a related note, is basecalling with buttery-eel using BLOW5 expected to be quicker than native Guppy/FAST5?

Thanks!

**Comment by Psy-Fer** [comment]
Hey,

I'm currently working my way through an update, so I can certainly make sure that `--chunks_per_runner` is one of the flags that can be accessed.

Any other options you would like to see while I'm at it?

As for performance, fast5/HDF5 is always going to be slower than slow5 if you are IO bound. So if you were using a fast model, slow5 is much faster than fast5. (even with only single threaded slow5 sequential access). Once you get to GPU bound models like HAC and SUP, it's more about the GPU you are running and how many. On the plus side, using this method with slow5 will populate all GPU resources, so that' helps, and then IO won't ever really be an issue either as you scale that.

So short answer, yes, but it depends, lol.

I hope this answers your question, and expect an update soon (a week or so) for the new options.

James

**Comment by mshamash** [comment]
Hi James,

Thanks for your quick reply!

`--chunks_per_runner` is probably the main flag I'd use in this context. I also use a lot of the demultiplexing and primer/adapter/barcode trimming features of guppy_basecaller, but I imagine that the buttery-eel wrapper was really just designed with basecalling in mind, and that demuxing would be done downstream with guppy_barcoder. Otherwise it quickly becomes a lot of flags to add.

Thanks for the performance info. We use the SUP model usually on our A100 GPUs, so I do think we are still GPU-bound in the end - so in that case I shouldn't expect any big performance gains by basecaling with blow5. However something I noticed when scaling up to 3x or 4x A100 is that I start to become IO bound. Could be interesting to compare fast5/blow5 on these multi-GPU systems. I'll probably do this sometime in the next week or so, and will add a new tab to my benchmark data [here](https://docs.google.com/spreadsheets/d/1eu49_jdRsDwnxwMzRCWAYpz5F6yHE8wm6xSs0OZpSvk/edit#gid=0) for comparison.

Michael

**Comment by hasindu2008** [comment]
@mshamash Looking at your results, I guess that BLOW5 will benefit the A100s significantly for FAST and HAC models and only slightly for SUP. 

What I observed on Tesla V100 GPUs on a spinning disk NAS mounted over NFS for HAC model (dataset had 500,000 reads):
| GPUs | FAST5 time (minutes) | BLOW5 time (minutes)  |
|------|------------------|-------------------|
| 1    | 69               | 59                |
| 2    | 64               | 32                |
| 3    | 64               | 23                |
| 4    | 65               | 20                |

These results are however based on BLOW5 directly integrated to Guppy at C++ level (which we have access to through dev license). I would expect buttery-eel to perform similarly at least closely.

**Comment by Psy-Fer** [comment]
hey @mshamash,

I've got a fresh update on the dev branch which allows for most options to be parsed to guppy. So if you give it `--chunks_per_runner` it should know exactly what you are doing and work with it. 

Options that would give a sam output will still only give a fastq output though. So no alignment or barcode demux options just yet, but working on that now.

When I've finished these other additions I'll cut a version on main, but thought i'd give you a chance to checkout a preview if you were keen.

here is the specific working commit (in case more recent commits make the dev branch unstable)

https://github.com/Psy-Fer/buttery-eel/commit/e23521363af082d14fcf9da8b8cee57718f64f9e

- James
